_drafts/gdb-python-post.md

```diff
@@ -0,0 +1,378 @@
+
+The GDB Python API is a GDB compile time option that can be enabled (with the `--with-python` configuration argument). We can easily check if gdb has this feature enabled by checking:
+
+```bash
+$ /path/to/gdb  --config
+This GDB was configured as follows:
+[...]
+             --with-python=/usr
+[...]
+```
+
+If you are doing any type of ARM Cortex-M development, you are in luck, the [official ARM Embedded Toolchain](https://developer.arm.com/tools-and-software/open-source-software/developer-tools/gnu-toolchain/gnu-rm/downloads) publishes a GDB variant with Python enabled at `./bin/arm-none-eabi-gdb-py` in each release!
+
+To use it you just need python installed on your system (i.e On ubuntu, `apt-get install python2.7 python-dev` or on OSX at the moment, the pre-installed system python is used)
+
+## Example Usage
+
+Let's walk through a trivial example snippet of code to get an idea of how some of the APIs work!
+
+The snippet tracks UUIDs in RAM in a linked list. On a real embedded system, we may have a bunch of items, identified by their UUID, stored on an external flash. If the list of items are looked at frequently, it may make sense for the list may be cached in RAM for faster lookup (since flash access can be slow) and better power (since reading from external flash consumes more power than a RAM read).
+
+```c
+#include <inttypes.h>
+#include <stddef.h>
+#include <stdlib.h>
+
+typedef struct Uuid {
+  uint8_t bytes[16];
+} Uuid;
+
+typedef struct UuidListNode {
+  struct UuidListNode *next;
+  Uuid uuid;
+} UuidListNode;
+
+static UuidListNode *s_list_head = NULL;
+
+void list_add_uuid(const Uuid *fake_uuid) {
+  UuidListNode *node_to_add = malloc(sizeof(UuidListNode));
+  if (node_to_add == NULL) {
+    return;
+  }
+
+  *node_to_add = (UuidListNode) {
+    .uuid = *fake_uuid,
+  };
+
+  // Add entry to the front of the list
+  node_to_add->next = s_list_head;
+  s_list_head = node_to_add;
+}
+
+//
+// Code to populate the list with fake data
+//
+
+// A completely fake uuid-generator implementation
+static void generate_fake_uuid(Uuid *uuid) {
+  for (size_t i = 0; i < sizeof(*uuid); i++) {
+    uuid->bytes[i] = rand() & 0xff;
+  }
+}
+
+static void generate_fake_uuid_list(void) {
+  for (int i = 0; i < 10; i++) {
+    Uuid uuid;
+    generate_fake_uuid(&uuid);
+    list_add_uuid(&uuid);
+  }
+}
+
+int main(void) {
+  generate_fake_uuid_list();
+  // Let's just spin loop so we can break at any point and dump the list
+  while (1) { }
+  return 0;
+}
+```
+
+### GDB Pretty Printers
+
+When we print a structure, GDB will by default will display each of the individual fields. The `s_list_head` in the code above would look something like this:
+
+```
+(gdb) p s_list_head
+$1 = (UuidListNode *) 0x200070c0
+(gdb) p *s_list_head
+$2 = {next = 0x200070a8, uuid = {bytes = "\235]D@\213Z#^\251\357\354?\221\234zA"}}
+```
+
+We can clean it up a little by enabling pretty printing which will put each field on a newline:
```

Comment:
Show the command to enable pretty printing in the code snippet below

Reply:
yeah, how do I enable pretty printing?

Reply:
I swear that was there at some point hehe, thanks!